dataverse-webapp/src/main/java/edu/harvard/iq/dataverse/datafile/file/FileDownloadAPIHandler.java

```diff
@@ -0,0 +1,444 @@
+                            // to produce some output.
+                            zipperWrapper.init(outputStream);
+                        }
+
+                        long size = computeFileSize(file, originalFileFormat);
+                        if (size < (zipDownloadSizeLimit - sizeTotal)) {
+                            sizeTotal += zipperWrapper.getZipper().addFileToZipStream(file, originalFileFormat);
+                            filesToDownload.add(file);
+                        } else {
+                            String fileName = file.getFileMetadata().getLabel();
+                            String mimeType = file.getContentType();
+                            zipperWrapper.addToManifest(fileName + " (" + mimeType +
+                                                                ") skipped because the total size of the download bundle exceeded the limit of "
+                                                                + zipDownloadSizeLimit + " bytes.\r\n");
+                        }
+
+                    }
+                }
+            }
+
+            if (zipperWrapper.isEmpty()) {
+                // If the DataFileZipper object is still NULL, it means that there were file ids supplied - but none of
+                // the corresponding files were accessible for this user.
+                // In which case we don't bother generating any output, and just give them a 403:
+                throw new ForbiddenException();
+            }
+
+            // Check whether some subset of downloaded files is equal to some whole
+            // set of files of some version and log if so
+            wholeDatasetDownloadLogger.incrementLogIfDownloadingWholeDataset(filesToDownload);
+
+            // This will add the generated File Manifest to the zipped output, then flush and close the stream:
+            zipperWrapper.getZipper().finalizeZipStream();
+        };
+    }
+
+    private long determineDownloadSizeLimit() {
+        long limit = settingsService.getValueForKeyAsLong(SettingsServiceBean.Key.ZipDownloadLimit);
+        if (limit == -1) {
+            throw new BadRequestException("Download zipped bundles of multiple files option is disabled in this installation");
+        }
+        return limit != 0 ? limit : Long.MAX_VALUE;
+    }
+
+    private long computeFileSize(DataFile file, boolean getOriginal) throws IOException {
+        long size;
+        // is the original format requested, and is this a tabular datafile, with a preserved original?
+        if (getOriginal && file.isTabularData() && StringUtil.nonEmpty(file.getDataTable().getOriginalFileFormat())) {
+
+            // We now store the size of the original file in the database (in DataTable), so we get it for free.
+            // However, there may still be legacy datatables for which the size is not saved. so the "inefficient" code
+            // is kept, below, as a fallback solution.
+            // -- L.A., 4.10
```

Comment:
Even though we may not have legacy 'datatables' (also we would have to know if datatable on production is also containing all the data), i still think it is a reasonable fallback since we don't have many proper tests I feel like defensive programming would help us in this case (since there is no reason why there wouldn't be such a bug in the future, with higher probability without effective tests).